File: mobile/lib/core/models/patient_profile.dart
enum PatientRelationship {
  me,
  mother,
  father,
  wife,
  husband,
  son,
  daughter,
  brother,
  sister,
  grandfather,
  grandmother,
  other,
}

enum PatientGender { female, male, unspecified }

class PatientProfileId {
  const PatientProfileId(this.value) : assert(value != '');
  static const me = PatientProfileId('me');
  final String value;
  @override
  bool operator ==(Object other) =>
      other is PatientProfileId && other.value == value;
  @override
  int get hashCode => value.hashCode;
}

class PatientProfile {
  const PatientProfile({
    required this.id,
    required this.relationship,
    required this.firstName,
    required this.lastName,
    required this.gender,
    required this.dateOfBirth,
  });
  final PatientProfileId id;
  final PatientRelationship relationship;
  final String firstName, lastName;
  final PatientGender gender;
  final DateTime dateOfBirth;
  String get displayName => '$firstName $lastName'.trim();
  String get initials =>
      '${firstName.isEmpty ? '' : firstName[0]}${lastName.isEmpty ? '' : lastName[0]}'
          .toUpperCase();
  int ageAt(DateTime date) {
    var age = date.year - dateOfBirth.year;
    if (date.month < dateOfBirth.month ||
        (date.month == dateOfBirth.month && date.day < dateOfBirth.day)) {
      age--;
    }
    return age;
  }
}

File: backend/src/modules/patients/domain/patient-profile.ts
export const patientRelationships = [
  'self',
  'mother',
  'father',
  'son',
  'daughter',
  'brother',
  'sister',
  'grandfather',
  'grandmother',
  'wife',
  'husband',
  'other',
] as const;
export type PatientRelationship = (typeof patientRelationships)[number];

export const patientGenders = ['female', 'male', 'unspecified'] as const;
export type PatientGender = (typeof patientGenders)[number];

export interface PatientProfileProjection {
  readonly id: string;
  readonly firstName: string;
  readonly lastName: string;
  readonly dateOfBirth: string;
  readonly gender: PatientGender;
  /** Flutter compatibility: backend `self` is exposed as `me`. */
  readonly relationship: Exclude<PatientRelationship, 'self'> | 'me';
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly active: boolean;
  readonly version: number;
}

export interface PatientAccountProjection {
  readonly id: string;
  readonly activeProfileId: string | null;
  readonly activeProfile: PatientProfileProjection | null;
  readonly profileCount: number;
  readonly createdAt: string;
  readonly updatedAt: string;
}

export function apiRelationship(
  value: PatientRelationship,
): PatientProfileProjection['relationship'] {
  return value === 'self' ? 'me' : value;
}
